tsdb/head.go

```diff
@@ -424,6 +438,31 @@ func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOpti
 
 func mmappedChunksDir(dir string) string { return filepath.Join(dir, "chunks_head") }
 
+func (h *Head) ApplyConfig(cfg *config.Config) error {
+	if h.opts.EnableExemplarStorage && cfg.StorageConfig.ExemplarsConfig.MaxExemplars > -1 {
+		// ugly hack to get the previous default size
+		if cfg.StorageConfig.ExemplarsConfig.MaxExemplars == 0 {
+			cfg.StorageConfig.ExemplarsConfig.MaxExemplars = 100000
+		}
+		switch h.exemplars.(type) {
+		case noopExemplarStorage:
+			e, err := NewCircularExemplarStorage(cfg.StorageConfig.ExemplarsConfig.MaxExemplars, h.reg)
+			if err != nil {
+				return err
+			}
+			h.exemplars = e
```

Comment:
This will cause race with the appending since there is nothing protecting this `ApplyConfig`/`h.exemplars`. I think we should get rid of `noopExemplarStorage` and just start with a 0 size exemplar storage in `NewHead` without having to lock `ApplyConfig`.

Reply:
Good point. We'd still be allocating *some* memory with a 0 length buffer and empty map, but I think that's reasonable.

I'll think about this a little before making any changes here.

Reply:
The race still exists

Reply:
I'll deal with this one in the morning